napari/layers/labels/labels.py

```diff
@@ -447,23 +447,35 @@ def fill(self, coord, old_label, new_label):
         ----------
         coord : sequence of float
             Position of mouse cursor in image coordinates.
-        old_label : int
-            Value of the label image at the coord to be replaced.
         new_label : int
             Value of the new label to be filled in.
+        refresh : bool
+            Whether to refresh view slice or not. Set to False to batch paint
+            calls.
         """
-        self._save_history()
-
-        int_coord = np.round(coord).astype(int)
+        if refresh is True:
+            self._save_history()
 
         if self.n_dimensional or self.ndim == 2:
             # work with entire image
             labels = self.data
-            slice_coord = tuple(int_coord)
+            slice_coord = tuple(
+                np.round(np.clip(c, 0, s - 1)).astype(int)
+                for c, s in zip(coord, self.shape)
+            )
         else:
             # work with just the sliced image
             labels = self._data_raw
-            slice_coord = tuple(int_coord[d] for d in self.dims.displayed)
+            slice_coord = tuple(
+                np.round(np.clip(coord[d], 0, self.shape[d] - 1)).astype(int)
+                for d in self.dims.displayed
+            )
+
+        self.coordinates = slice_coord
+
+        old_label = self.get_value()
+        if old_label == self.selected_label:
```

Comment:
```suggestion
        if old_label == new_label:
```
this check should be against `new_label` which is going to be filled in